skimage/feature/sift.py

```diff
@@ -0,0 +1,721 @@
+           vol. 60, pp. 91–110.
+           :DOI:`10.1023/B:VISI.0000029664.99615.94`
+
+    .. [3] I. R. Otero and M. Delbracio. "Anatomy of the SIFT Method",
+           Image Processing On Line, 4 (2014), pp. 370–396.
+           :DOI:`10.5201/ipol.2014.82`
+
+    Examples
+    --------
+    >>> from skimage.feature import SIFT, match_descriptors
+    >>> from skimage.data import camera
+    >>> from skimage.transform import rotate
+    >>> img1 = camera()
+    >>> img2 = rotate(camera(), 90)
+    >>> detector_extractor1 = SIFT()
+    >>> detector_extractor2 = SIFT()
+    >>> detector_extractor1.detect_and_extract(img1)
+    >>> detector_extractor2.detect_and_extract(img2)
+    >>> matches = match_descriptors(detector_extractor1.descriptors,
+    ...                             detector_extractor2.descriptors,
+    ...                             max_ratio=0.6)
+    >>> matches[10:15]
+    array([[ 10, 412],
+           [ 11, 417],
+           [ 12, 407],
+           [ 13, 411],
+           [ 14, 406]])
+    >>> detector_extractor1.keypoints[matches[10:15, 0]]
+    array([[ 95, 214],
+           [ 97, 211],
+           [ 97, 218],
+           [102, 215],
+           [104, 218]])
+    >>> detector_extractor2.keypoints[matches[10:15, 1]]
+    array([[297,  95],
+           [301,  97],
+           [294,  97],
+           [297, 102],
+           [293, 104]])
+
+    """
+
+    def __init__(self, upsampling=2, n_octaves=8, n_scales=3, sigma_min=1.6,
+                 sigma_in=0.5,
+                 c_dog=0.04 / 3, c_edge=10, n_bins=36, lambda_ori=1.5,
+                 c_max=0.8, lambda_descr=6, n_hist=4, n_ori=8):
+        if upsampling in [1, 2, 4]:
+            self.upsampling = upsampling
+        else:
+            raise ValueError("upsampling must be 1, 2 or 4")
+        self.n_octaves = n_octaves
+        self.n_scales = n_scales
+        self.sigma_min = sigma_min / upsampling
+        self.sigma_in = sigma_in
+        self.c_dog = (2 ** (1 / n_scales) - 1) / (2 ** (1 / 3) - 1) * c_dog
+        self.c_edge = c_edge
+        self.n_bins = n_bins
+        self.lambda_ori = lambda_ori
+        self.c_max = c_max
+        self.lambda_descr = lambda_descr
+        self.n_hist = n_hist
+        self.n_ori = n_ori
+        self.delta_min = 1 / upsampling
+        self.float_dtype = None
+        self.scalespace_sigmas = None
+        self.keypoints = None
+        self.positions = None
+        self.sigmas = None
+        self.scales = None
+        self.orientations = None
+        self.octaves = None
+        self.descriptors = None
+
+    @property
+    def deltas(self):
+        """The sampling distances of all octaves"""
+        deltas = self.delta_min * np.power(2, np.arange(self.n_octaves),
+                                           dtype=self.float_dtype)
+        return deltas
+
+    def _set_number_of_octaves(self, image_shape):
+        size_min = 12  # minimum size of last octave
```

Comment:
Shouldn't `size_min` be a parameter here?

Reply:
I thought about that too. In the original implementation this is a constant and can't be changed. My reason not to do it was that the user can still practically increase size_min by lowering n_octaves and I'm not sure if it makes sense to lower size_min even more.
But if you prefer it as a parameter I have no problem with that. We just need to find some minimal value, because size_min too low would probably start to break some things.

Reply:
An other option would be to make it private attribute of the `SIFT` class to make it twickable for "expert" users...

Reply:
That sounds good, I will add it.

Reply:
I plan to merge this PR soon for inclusion in the 0.19 release. If this does not get addressed before then, it can go in a follow-up PR as it would not be a public API change.

Reply:
Let's leave it as is for now.  We can always add this feature if it is requested, but if it is not a parameter in the original it seems unlikely.  Also, I am not enthusiastic about private attributes tweakable by expert users: who should either have an API for it, or it should not be adjustable.